### PR TITLE
Split visualizations into packs

### PR DESCRIPTION
**Pull request by axelboc**
Split out NeXus visualizations from core visualizations by introducing the concept of "visualization packs".

image

`App` now renders a component called `VisPackChooser`, which fetches the entity at the selected path and determines which visualization pack to render. If the entity has a `default` attribute or is an `NXdata` group, then `NexusPack` is rendered; otherwise `CorePack` is rendered.

`CorePack` and `NexusPack` can then determine which of their visualizations are supported by the entity they receive, in which ever way they see fit. They then render the `Visualizer` component, which receives an `activeVis` definition object (with the `Icon`, `Container`, etc. to use).

### CODE REVIEW

**Review by axelboc — COMMENTED** [review]
I feel like this was worthwhile:

- The NeXus visualization are now more clearly separate from the core visualization in the tree and decoupled from them in the code.
- Letting each pack check which of its visualizations are supported by a given entity is very flexible, and will allow us to more easily fetch entities from the store lazily if need be.
- Almost no duplication was introduced thanks to the sharing of `Visualizer` and the new `useActiveVis` custom hook. We've actually removed duplication by combining the two Nexus `supportsEntity` functions directly into the Nexus pack's `getSupportedVis` function.

**Comment by axelboc** [comment]
You should probably pull the branch to get a better feel for the new structure. I'm not sure about the location of a couple of files, like `VisPackChooser`.

**Review by loichuder — CHANGES_REQUESTED** [review]
I would put `VisPackChooser` and `visualizer/hooks` at the root of `vis-packs`.

Also, it feels weird to have `models` files in a `shared` folder and one in the root of the pack folder. What about having the `shared` folder only contain components while other files (`models`, `utils` and `hooks`) are moved at the root of the pack folder ?

`Visualizer` name is no longer fitting in my opinion. It can stay like this for now but we should perhaps think about another name (we could revert to `VisDisplay` ?)

File: src/h5web/vis-packs/nexus/shared/utils.test.ts [hunk 1]
```diff
@@ -108,17 +107,6 @@ describe('NeXus utilities', () => {
...
-  describe('isNxInterpretation', () => {
-    it('should return `true` with known NeXus interpretation', () => {
-      expect(isNxInterpretation('spectrum')).toBe(true);
-    });
-
-    it('should return `false` with invalid or unknown NeXus interpretation', () => {
-      expect(isNxInterpretation(42)).toBe(false);
-      expect(isNxInterpretation('unknown')).toBe(false);
-    });
-  });
```
[the remark this answers is not in the document]
Comment (axelboc) [reply src/h5web/vis-packs/nexus/shared/utils.test.ts:120]: `isNxInterpretation` is no longer an exported utility -- it's used by the Nexus packs' `getSupportedVis` utility and is therefore tested by this utility's tests.

File: src/h5web/vis-packs/nexus/utils.test.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+const datasetFlt3D = makeSimpleDataset('dataset_flt_3d', intType, [5, 3, 1]);
+
+describe('Nexus pack utilities', () => {
+  describe('getSupportedVis', () => {
+    it('should not include NxSpectrum vis if any other visualization is supported', () => {
+      const datasetInt2D = makeSimpleDataset('dataset', intType, [5, 3]);
+      const nxDataSignal2D = makeNxDataGroup('foo', { signal: datasetInt2D });
+      const supportedVis = getSupportedVis(nxDataSignal2D);
+
+      expect(supportedVis).toEqual(NEXUS_VIS.image);
```
[the remark this answers is not in the document]
Comment (axelboc) [reply src/h5web/vis-packs/nexus/utils.test.ts:26]: I'm just as surprise as you, but this is a valid way to access a value in a `Record` indexed by an enum. It autocompletes and everything!

File: src/h5web/visualizer/VisPackChooser.tsx [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  const entity = entitiesStore.get(path);
+
+  if (
+    getAttributeValue(entity, 'default') ||
+    getAttributeValue(entity, 'NX_class') === 'NXdata'
+  ) {
+    return <NexusPack entity={entity} />;
+  }
+
+  return <CorePack entity={entity} />;
```
[the remark this answers is not in the document]
Comment (axelboc) [reply src/h5web/visualizer/VisPackChooser.tsx:24]: `VisPackChooser`? The logic is covered by the E2E specs. I don't think integration tests would add any value.

**Comment by axelboc** [comment]
> I would put VisPackChooser and visualizer/hooks at the root of vis-packs.

Yeah, I wasn't sure because you could argue that the `Visualizer` itself is also a shared component of the two packs. Since we extracted it into its own root folder, I thought maybe the other things that were shared by the two packs had to go in the same folder, but no, they're different things, you're right. 👍 

> Also, it feels weird to have models files in a shared folder and one in the root of the pack folder. What about having the shared folder only contain components while other files (models, utils and hooks) are moved at the root of the pack folder ?

Yeah, the problem was differentiating the pack's utilities (`getSupportedVis`, `getDefaultPath`), from the visualization's utilities (which are numerous). Perhaps we could rename the files `pack-utils` and `vis-utils` (or keep the latter as `utils`)? 

> Visualizer name is no longer fitting in my opinion. It can stay like this for now but we should perhaps think about another name (we could revert to VisDisplay ?)

True. It's a shame that `vis-display` wouldn't be a great name for the folder...

**Comment by loichuder** [comment]
> Yeah, the problem was differentiating the pack's utilities (`getSupportedVis`, `getDefaultPath`), from the visualization's utilities (which are numerous). Perhaps we could rename the files `pack-utils` and `vis-utils` (or keep the latter as `utils`)?

`pack-utils` and `utils` seems good to me.

**Comment by axelboc** [comment]
Cool, moving the utility/model/hook files out of the `shared` folders really cleans things up. 💯 

- I was able to merge two `hooks` files in the core pack (`core/hooks` which contained `useDatasetValue/s` and `core/shared/hooks`).
- I moved the raw and scalar vis and their CSS files in sub-folders to match the other core visualisations.
- I moved the Spectrum vis toolbar config from `nexus/shared/config` to `nexus/spectrum/config`. It's the only file in the folder, but it matches the structure inside the core pack.

**Review by loichuder — APPROVED** [review]
Looks good to me !

**Pull request merged**